src/lib/tls/msg_client_hello.cpp

```diff
@@ -513,12 +531,22 @@ Client_Hello_12::Client_Hello_12(Handshake_IO& io,
    if(policy.support_cert_status_message())
       m_data->extensions().add(new Certificate_Status_Request({}, {}));
 
-   auto supported_groups = std::make_unique<Supported_Groups>(policy.key_exchange_groups());
-   if(!supported_groups->ec_groups().empty())
+   // RFC 8446 4.2.3
+   //    The client MUST NOT include these extensions [Supported Elliptic Curves Extension
+   //    and Supported Point Formats Extension] in the ClientHello message if it does not
+   //    propose any ECC cipher suites.
+   auto offered_ciphers = m_data->ciphersuites();
+   if (std::any_of(offered_ciphers.begin(), offered_ciphers.end(), [](uint16_t cipherSuiteCode) {
+      std::optional<Ciphersuite> cs = Ciphersuite::by_id(cipherSuiteCode);
+      return cs.has_value() && (cs.value().ecc_ciphersuite()); }))
       {
-      m_data->extensions().add(new Supported_Point_Formats(policy.use_ecc_point_compression()));
+      auto supported_groups = std::make_unique<Supported_Groups>(policy.key_exchange_groups());
+      if(!supported_groups->ec_groups().empty())
+         {
+         m_data->extensions().add(new Supported_Point_Formats(policy.use_ecc_point_compression()));
+         }
```

Comment:
Question: Is it possible/legal that we might offer some ECC-based ciphersuites (outer `if` statement is true) but don't advertise any supported ECC groups (this `if` statement is also true)?

Reply:
I think technically it is allowed, but wouldn't make sense to do since the ECC suites could not be negotiated.RFC 8422 states that if impl is "unable to complete the ECC handshake while restricting itself to the enumerated curves and point formats, it MUST NOT negotiate the use of an ECC cipher suite." and TLS 1.3 has similar language.

Reply:
I simply copied the previous code and put it in the new if clause. I think an additional check can't hurt, but I'll remove it to keep it more simple.